### ISSUE TITLE
[FEA][AUDIT][SPARK-48949][SQL] SPJ: Runtime partition filtering

### ISSUE THREAD

**Issue by abellina** [NVIDIA/cudf-spark#11453]
https://github.com/apache/spark/commit/dbba92a1b4e adds partition filtering for inner, left, and right outer joins at the scan level. We have part of this code in `GpuBatchScanExec` and need to port.

**Comment by amahussein** [comment]
## This is a latent bug, not only a missed optimization

With `spark.sql.sources.v2.bucketing.partition.filter.enabled=true` **and** `spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled=true`, an SPJ inner join over GPU Iceberg scans fails on the driver before any task is scheduled:

```
java.lang.AssertionError: assertion failed: Partition value ... does not exist in
  common partition values from Spark plan
  at com.nvidia.spark.rapids.shims.GpuBatchScanExec.$anonfun$inputRDD$11(GpuBatchScanExec.scala:207)
```

Reproduced on `buildver=402`, Spark 4.0.2, Iceberg 1.10.1, one RTX A5000. CPU Spark 4.0.2 runs the identical query fine, so this is ours. SPARK-48949 made `commonPartitionValues` an *intersection* where the plugin's copy of `BatchScanExec.inputRDD` still assumes a *superset*; the `filteredGroupedPartitions` guard upstream added to drop the scan's extra groups was never copied across.

Full matrix, with join-stage RDD partition counts:

| `partition.filter` | `partiallyClustered` | GPU |
|---|---|---|
| off | off | works (3) |
| off | on  | works (4) |
| on  | off | works, **already prunes** (2) |
| on  | on  | **`AssertionError`** |

Worth stating plainly because it changes how this should be sold: row 3 shows the pruning benefit is **already delivered** on the path that works today, with no plugin change. This port removes a crash and extends pruning to the one path that crashes. It is not a throughput win.

## Scope of the fix

- The `filteredGroupedPartitions` guard in `sql-plugin/src/main/spark350db143/.../GpuBatchScanExec.scala` (serves `350db143`, `400`, `400db173`, `401`–`404`, `411`–`413`) and in `spark340/.../GpuBatchScanExec.scala` (serves 340–344, 350–359). The `spark340` copy is **inert** — the config does not exist before Spark 4.0 and `dbba92a1b4e` is not on `branch-3.5` — but is included so the two copies stay structurally identical and the next audit diff stays cheap.
- A new Iceberg integration test. It will **not** extend `test_iceberg_spj_partial_clustering_distinct`: that test's 3.5.9+/4.0.3+/4.1.2+ gate exists only because it asserts on `SELECT DISTINCT`, which SPARK-55848 breaks on CPU as well as GPU. Asserting on join rows instead needs no patch-level gate, so a new test actually runs on the distros we build against.
- A performance statement, since upstream (apache/spark#47426) supplied no numbers to inherit.

## Deferred: SPARK-58783

`filteredPartitions` in both shims still reads `val originalPartitioning = outputPartitioning` where upstream now has `super.outputPartitioning` ([SPARK-58783](https://issues.apache.org/jira/browse/SPARK-58783), `c995c434117`). I reproduced its failure mode (`SparkException: During runtime filtering, data source must not report new partition values that are not present in the original partitioning`) on the same setup.

**Deferring it, deliberately.** CPU Spark 4.0.2 fails identically without the plugin, and no released Spark carries the fix — the newest `apache/spark` tags are v4.0.4 and v4.1.3, and our buildvers stop at `404`/`413`. So there is no version-specific exposure today, and porting it now would make the plugin succeed where every Spark we support throws, which no `assert_gpu_and_cpu_are_equal_collect` test can express. Tracked in #15839, gated on a 4.0.5 or 4.1.4 shim existing.

## Also out of scope, tracked in #15338

While scoping the above I hit a second, independent divergence in the same file and confirmed it by running it. It is already filed as #15338, so the reproduction is recorded there.

`GpuBatchScanExec.outputPartitioning` only rewrites the partitioning when `spjParams.commonPartitionValues` is defined, while Spark also projects onto the join keys when only `spjParams.joinKeyPositions` is defined. On a one-sided SPJ the GPU scan then reports 4 partitions where Spark reports 3, and the query dies with `IllegalArgumentException: Can't zip RDDs with unequal numbers of partitions: List(4, 3)`; CPU returns correct results. Reached via `allowJoinKeysSubsetOfPartitionKeys.enabled` + `v2.bucketing.shuffle.enabled`, so it cannot co-occur with this issue's configs. Separate PR.

## Considered and set aside: delegating instead of copying

Both full-SPJ shims are hand-copies of Spark internals that have now drifted in **five** places — `reducers` (#15721, fixed), `hashCode` (#15666, fixed), SPARK-48949 (this issue), the `outputPartitioning` projection (#15338) and SPARK-58783 (#15839). So it is fair to ask whether to stop copying. Per tier:

- **Spark 4.0 — not available.** The logic is inline in `BatchScanExec.inputRDD` with no trait or helper to call, and `GpuBatchScanExec` extends `GpuBatchScanExecBase`, not Spark's `BatchScanExec`. Spark 4.2's `GroupPartitionsExec` model is not portable backwards either — that node is inserted by Spark's own `EnsureRequirements`.
- **Spark 4.1 — available.** SPARK-53272 moved the logic into `org.apache.spark.sql.execution.KeyGroupedPartitionedScan`, a public trait with no abstract members, so `411`–`413` could mix it in and drop ~125 lines. It helps none of the seven 4.0-family buildvers the same file serves, so it is a separate change, not part of this fix — tracked in #15840, to be done after this lands so the refactor has a test protecting it.

So the copy stays on 4.0 regardless, and this fix is not throwaway under either outcome.
